### PR TITLE
Use Conv1d for TDNN

### LINKED ISSUE #25476: Some problems with XVector implementation
### System Info

NA

### Who can help?

@sanchi

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

While working on #25471, I spot a few problems with the current XVector implementation

1. Statistical pooling is using for loops for batched computation 

https://github.com/huggingface/transformers/blob/fe3c8ab1af558b95f67f5fafc0c55f09fd2b09db/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L2436-L2438

This can be avoided by using the attention mask. However, since TDNN layers (basically 1D-conv) also change the effective attention mask, we need to update the attention mask again.

2. Std uses unbiased estimate (correction=1), while from what I know, statistical pooling typically uses biased estimate.

See that PyTorch uses correction=1 by default (unbiased=True in previous versions): https://pytorch.org/docs/stable/generated/torch.std.html

The original papers (x-vector: https://www.danielpovey.com/files/2018_icassp_xvectors.pdf, stats pooling: http://danielpovey.com/files/2017_interspeech_embeddings.pdf) don't specify the equations, but most other papers show equations without the correction factor: https://arxiv.org/pdf/1803.10963.pdf.

The official implementation is included in Kaldi, but I don't understand C++ and their codebase well enough to understand what's going on. https://github.com/kaldi-asr/kaldi/blob/71f38e62cad01c3078555bfe78d0f3a527422d75/src/nnet3/nnet-general-component.cc#L808-L821

If we use HF to train a model, it wouldn't matter much. But if we use this to port weights from other implementation, this can be a problem.

3. TDNN layer is 1D-conv. Using unfold and linear is wasteful. We can directly use 1D-conv instead.

NeMo's TDNN: https://github.com/NVIDIA/NeMo/blob/ab749e4401a3dcdfa5ea969347aaee20b7947c7c/nemo/collections/asr/parts/submodules/tdnn_attention.py#L138

### Expected behavior

NA

### PR DESCRIPTION
**Pull request by gau-nernst**
# What does this PR do?

Partially fixes #25476 

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by ArthurZucker** [comment]
cc @sanchit-gandhi

File: src/transformers/models/data2vec/modeling_data2vec_audio.py
Comment (sanchit-gandhi) [inline src/transformers/models/data2vec/modeling_data2vec_audio.py:1370]: Very clean, like this solution a lot @gau-nernst! Do you know what kind of speed-ups we get already?

Comment (gau-nernst) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1370]: I will try to run some benchmarks and let you know.
Since I'm also working on LoRA at my work, I realize doing this will bypass any "magic" to be applied on nn.Linear, since we don't call nn.Linear.forward(). Although LoRA is not likedly to be applied here, since the TDNN layers will most likely be trained from scratch, it may lead to unexpected behavior if the user wants to "hack" nn.Linear.
Another solution is to replace nn.Linear completely with nn.Conv1d, and apply some load_state_dict hook for backward compatability. But I think PyTorch's load state dict hook is only available recently. I will check on this.

**Comment by gau-nernst** [comment]
On RTX 3090, 5s input

dtype | batch size | type | de139702a | 4ad9e29a5
--|--|--|--|--
torch.float32 | 1 | forward | 61.8698 it/s | 65.4886 it/s
torch.float32 | 1 | forward+backward | 15.7255 it/s | 15.3732 it/s
torch.float32 | 8 | forward | 12.1235 it/s | 13.3573 it/s
torch.float32 | 8 | forward+backward | 4.3197 it/s | 4.4702 it/s
torch.float16 | 1 | forward | 70.7141 it/s | 73.2518 it/s
torch.float16 | 1 | forward+backward | 17.1505 it/s | 17.1628 it/s
torch.float16 | 8 | forward | 20.3633 it/s | 24.0503 it/s
torch.float16 | 8 | forward+backward | 7.5402 it/s | 8.0300 it/s

The improvement is around 5%. It's small but not insignificant.

Another optimization we can do is to make `TDNNLayer` accepts channels-first input instead of channels-last (current approach) to avoid transposing data twice every layer. However, this would break backward compatibility if someone relies on `TDNNLayer` externally (though I think it's unlikely).

Benchmark script

```python
import time
from itertools import product
import torch
from transformers.models.wav2vec2.modeling_wav2vec2 import Wav2Vec2ForXVector

torch.backends.cuda.matmul.allow_tf32 = True
torch.backends.cudnn.allow_tf32 = True
torch.backends.cudnn.benchmark = True

device = "cuda"
batch_sizes = (1, 8)
dtypes = (torch.float32, torch.float16)

m = Wav2Vec2ForXVector.from_pretrained("facebook/wav2vec2-xls-r-300m").to(device)

for dtype, bsize in product(dtypes, batch_sizes):
    m.to(dtype)
    x = torch.randn(bsize, 16_000 * 5, device=device, dtype=dtype)

    m.eval()
    with torch.no_grad():
        m(x)  # warmup
    torch.cuda.synchronize()

    N = 100

    with torch.no_grad():
        time0 = time.perf_counter()
        for _ in range(N):
            out = m(x)
            torch.cuda.synchronize()
    print(f"{dtype}, {bsize=}, forward: {N / (time.perf_counter() - time0):.4f} it/s")

    m.train()
    m(x)[0].sum().backward()  # warmup

    time0 = time.perf_counter()
    for _ in range(N):
        out = m(x)
        out[0].sum().backward()
        torch.cuda.synchronize()
    print(f"{dtype}, {bsize=}, forward+backward: {N / (time.perf_counter() - time0):.4f} it/s")
```

**Comment by sanchit-gandhi** [comment]
Very cool! Thanks for the results @gau-nernst and nice benchmark script! I think it's worth pursuing in this case: 5% is the lower end of what we'd expect to get for `torch.compile`, and this has proven to be quite high demand for optimising PyTorch models, so a 5% improvement gain would be welcomed by the Wav2Vec2 community

I would be against adding a breaking change with the channel-last approach, since this is in contradiction to the scope of the PR where we set out to optimise in an entirely non-breaking way

**Comment by gau-nernst** [comment]
Cool! Anything else you want me to add to the PR, perhaps apart from removing the commented out old code? Is there a correctness test that we can add?

I didn't benchmark with torch.compile() since Wav2Vec2 did not work with it yet (I hope the relevant PR will be merged soon). Would you want to see benchmark results with torch.compile() for TDNNLayer alone? Perhaps it can even optimize the old way (unfold + linear)

**Comment by sanchit-gandhi** [comment]
It looks super clean already @gau-nernst - I think all that's left to do is:
1. Remove the commented out code
2. Run the slow tests to check we still have equivalence (this will check for correctness): https://github.com/huggingface/transformers/blob/1c6f072db0c17c7d82bb0d3b7529d57ebc9a0f2f/tests/models/wav2vec2/test_modeling_wav2vec2.py#L1922

No worries about benchmarking with torch compile - the results you provided previously were more than enough to justify this code change. We won't expect all users to be using PT 2.0, so this speed-up will definitely hold in many cases

**Review by sanchit-gandhi — APPROVED** [review]
Thanks @gau-nernst! Really elegant solution 👏 Nice job on bringing the speed-ups without breaking changes

If the slow test mentioned before passes, then this is good to merge for me! Requesting final review from @ArthurZucker

**Review by ArthurZucker — COMMENTED** [review]
Very elegant! 
Regarding this:
> Another optimization we can do is to make TDNNLayer accepts channels-first input instead of channels-last (current approach) to avoid transposing data twice every layer. However, this would break backward compatibility if someone relies on TDNNLayer externally (though I think it's unlikely).

TDNN Is not documented so not really part of our Public API, which means we could change it (but mostly only if we really have a significant gain in performance)
It's a breaking change for people who trained a model with LoRA no? That my only concern

**Comment by gau-nernst** [comment]
I think it's unlikely that someone will apply LoRA on the TDNN layers. Most likely they will apply LoRA on the transformer backbone only, and train the TDNN layers from scratch.

**Comment by ArthurZucker** [comment]
My question is rather, what if someone has already apply LoRA on the TDNN layers? (Though it might not be recommended, someone who does not specifically know the model could have just quantized all the linear layers no?) 😉

**Comment by gau-nernst** [comment]
I can't answer that, though I doubt that people who know how to use LoRA, and use it on TDNN layers specifically (most LoRA guides only show how to use it with transformers as far as I know), will face problems with this change. You guys from HF should decide whether this outweighs the cost i.e. speed-up vs annoy people who use LoRA on TDNN layers, and what kind of guarantee the library should provide i.e. a model surgery technique (e.g. LoRA) should still work after an update.

**Comment by ArthurZucker** [comment]
Pretty sure it will break inference, but we can probably have some kind of fix / deprecation cycle for this 😓

File: src/transformers/models/data2vec/modeling_data2vec_audio.py
Comment (sanchit-gandhi) [inline src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: cc @younesbelkada as well: currently we have un-optimised TDNN code, where we call `nn.functional.unfold` to do a conv1d, then apply a linear layer `self.kernel`

@gau-nernst has proposed is a solution where we still use the weights of the linear `self.kernel` layer (rather than switching to a conv1d layer which would change the state dict), but then use `nn.functional.conv1d` to do the convolution

=> the benchmarks show there's about a 10% performance gain by doing this

However, we worry that since we don't call `self.kernel.forward`, and LoRA layers for this layer will no longer work for inference. Is this hunch correct?

Comment (younesbelkada) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: Correct, these changes will make LoRA unusable for the kernel weights on PEFT side as we call `torch.nn.functional.conv1d` instead of the forward of the module itself.
However, usually we only adapt self attention modules (query, key, value) and in rare cases MLP layers of the transformer block. So I am not sure users really adapt `kernel` layers. 
I am not sure if we could make that PR BC (through a config variable maybe?)

cc @pacman100 & @BenjaminBossan as well

Comment (BenjaminBossan) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: If this is an edge case that is unlikely to be practically used, I would be fine with breaking backwards compatibility here, given the benefits.

If we are worried about this, however, we could make a check and raise an error/give a warning if a user happens to apply LoRA here. That check could be made here (check if `self.kernel` looks like a LoRA layer) or we could add a check in PEFT if we detect that a user tries to apply LoRA to this specific layer of this specific model.

Btw. this also affects other adapters such as IA³.

Comment (sanchit-gandhi) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: Awesome - thanks both for your input! It sounds like it's super low likelihood that a user is applying LoRA to these TDNN kernel layers, so I'd be happy proceeding with these changes if it's good for you @gau-nernst @ArthurZucker?

If not, and you'd like to preserve backwards comp for this edge case, I like @BenjaminBossan's idea of checking whether the weights look like a LoRA layer, and triggering a warning if so. There's also the option of having **two** implementations for the TDNN layer:
1. Old, un-optimised variant that is LoRA compatible
2. New, optimised variant that is not LoRA compatible

=> we could use 2 by default, but switch to 1 if the user has applied LoRA to the kernel

Comment (ArthurZucker) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: I'm fine with just warning the user if there are lora weights !

Comment (sanchit-gandhi) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: Let's do that! Are you happy to proceed with the proposed changes @gau-nernst and throw a warning if the user is using LoRA weights?

Comment (gau-nernst) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: Sure. I have some ideas about how to detect LoRA
1. Check if the linear layer is a subclass of PEFT's `lora.LoraLayer` (or `lora.Linear`). This requires importing PEFT (we can skip if there is no PEFT installed). This won't be able to detect other LoRA implementations. `torch.compile()` is likely to still work.
2. Check if the linear layer has attribute `lora_A` and/or `lora_B`. This assumes the LoRA implementation uses those names. This approach can potentially detect more LoRA implementations, apart from PEFT. However, iirc, `getattr()` will create a graph break in `torch.compile()`.

What are your thoughts?

Comment (BenjaminBossan) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: > However, iirc, `getattr()` will create a graph break in `torch.compile()`.

Interesting, I didn't know that. Would the same happen when using `try: layer.lora_A; except AttributeError: ...`? Because I would prefer the 2nd version, so it would be unfortunate if it cannot work with `torch.compile`. It is extremely unlikely that we would change the attribute name in PEFT so this check would be robust.

Comment (sanchit-gandhi) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: Fine for me to do number 1 since probably most users will be using PEFT with `transformers`. Checking for weight names `lora_A` and/or `lora_B` is a bit more ad-hoc. So even if it might catch more cases, it's prone to missing / breaking with later versions of LoRA when these vars might change.

Comment (gau-nernst) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: Sorry for the long overdue update. I added the PEFT LoRA check then realized PEFT LoRA does not even work with Wav2Vec2 😂.

```python
from transformers.models.wav2vec2 import Wav2Vec2Model
from peft import LoraConfig, get_peft_model

model = Wav2Vec2Model.from_pretrained("facebook/wav2vec2-base")

config = LoraConfig(target_modules=["q_proj"])
lora_model = get_peft_model(model, config)
```

Error message

```
/Users/xxx/github/transformers/src/transformers/configuration_utils.py:380: UserWarning: Passing `gradient_checkpointing` to a config initialization is deprecated and will be removed in v5 Transformers. Using `model.gradient_checkpointing_enable()` instead, or if you are using the `Trainer` API, pass `gradient_checkpointing=True` in your `TrainingArguments`.
  warnings.warn(
Traceback (most recent call last):
  File "/Users/xxx/github/transformers/debug.py", line 7, in <module>
    lora_model = get_peft_model(model, config)
  File "/Users/xxx/miniforge3/envs/pytorch/lib/python3.10/site-packages/peft/mapping.py", line 103, in get_peft_model
    return PeftModel(model, peft_config, adapter_name=adapter_name)
  File "/Users/xxx/miniforge3/envs/pytorch/lib/python3.10/site-packages/peft/peft_model.py", line 119, in __init__
    model = self._prepare_model_for_gradient_checkpointing(model)
  File "/Users/xxx/miniforge3/envs/pytorch/lib/python3.10/site-packages/peft/peft_model.py", line 331, in _prepare_model_for_gradient_checkpointing
    model.enable_input_require_grads()
  File "/Users/xxx/github/transformers/src/transformers/modeling_utils.py", line 1333, in enable_input_require_grads
    self._require_grads_hook = self.get_input_embeddings().register_forward_hook(make_inputs_require_grads)
  File "/Users/xxx/github/transformers/src/transformers/modeling_utils.py", line 1352, in get_input_embeddings
    raise NotImplementedError
NotImplementedError
```

where should I open an issue for this? transformers or peft?

Comment (gau-nernst) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: The error above is due to `gradient_checkpointing=True` in `facebook/wav2vec2-base` config. Setting it to False will resolve the issue. Still, it is indeed a bug i.e. Wav2Vec2 + gradient_checkpointing + PEFT LoRA -> fail. Please direct me where to open an issue for this.

I tested `torch.compile()` and found out that it does not support warning in the `forward()` method.

```python
import torch
import warnings

class DummyModel(torch.nn.Module):
    def forward(self, x):
        warnings.warn("Make a warning")
        return x
    
m = torch.compile(DummyModel(), fullgraph=True)(torch.randn(2, 10))
```

Therefore, having a run-time check and warning will always break `torch.compile` graph.

Comment (sanchit-gandhi) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: Does this 'bug' happen on `Wav2Vec2ForCTC` or `Wav2Vec2ForXVector` as well? Note that `Wav2Vec2Model` is just the base model, so there's not really any notion of exporting it to PEFT for fine-tuning.

Comment (gau-nernst) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: It does happen to `Wav2Vec2ForXVector`. I used `Wav2Vec2Model` in the above example to isolate the problem to the base model, to make sure it is not from the XVector components.

Comment (sanchit-gandhi) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: Okay so the problem here is that Wav2Vec2 doesn't have input embeddings: it works directly with the raw audio waveform and computes positional embeddings directly from the input. Since there is no notion of input embeddings, we don't implement the method `.get_input_embeddings`. This breaks the PEFT LoRA logic, which expects this method to be implemented.

This suggests that Wav2Vec2 is not compatible at all with PEFT LoRA as things currently stand. Would you mind indeed opening an issue on PEFT with the codesnippet you shared, and a copy-and-paste explanation of why it doesn't work?

This means though that we don't need to worry about breaking LoRA for Wav2Vec2, since it doesn't work with it anyway at the moment.

Comment (gau-nernst) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: The error is due to setting `gradient_checkpointing=True` in Wav2Vec2, which is the default for `facebook/wav2vec2-base`. Other Wav2Vec2 checkpoints, like `facebook/wav2vec2-xls-r-300m`, have `gradient_checkpointing=False`, thus there is no error, and PEFT LoRA still works with them. Overriding `gradient_checkpointing=False` can also make `facebook/wav2vec2-base` works with PEFT LoRA.

Comment (sanchit-gandhi) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: Ah ok thanks for the clarification! Let's stick to our original strategy then of raising a warning if there are LoRA weights

Comment (sanchit-gandhi) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: Hey @gau-nernst - would you be interested in finishing this PR? Now that we've agreed on the logic here and you've done a great job at laying down the foundations, it should be quite fast to tidy it up and complete the feature!

Comment (gau-nernst) [reply src/transformers/models/data2vec/modeling_data2vec_audio.py:1368]: Hello @sanchit-gandhi. I'm happy to finish this. Do let me know which parts you want to be changed.

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks for adding the LoRA warnings, this PR looks great @gau-nernst 🙌 Requesting final review from @ArthurZucker!

**Review by sanchit-gandhi — APPROVED** [review]

**Review by ArthurZucker — APPROVED** [review]
Really sorry for the late review! 
Can you rebase and I'll merge!

**Comment by ArthurZucker** [comment]
Great work here on finding an elegant solution 🤗

**Comment by gau-nernst** [comment]
@ArthurZucker i have merged master and make some small changes accordingly

**Comment by ArthurZucker** [comment]
Thanks! Mergin 🤗

**Pull request merged**
